objax/nn/layers.py

```diff
@@ -327,6 +327,63 @@ def __call__(self, x: JaxArray) -> JaxArray:
         self.avg.value += (self.avg.value - x) * (self.momentum - 1)
         return self.avg.value
 
+class RNN(Module):
+    """ Recurrent Neural Network (RNN) block."""
+
+    def __init__(self,
+                 nstate: int,
+                 nin: int,
+                 nout: int,
+                 activation: Callable = jn.tanh,
+                 w_init: Callable = kaiming_normal):
+        """Creates an RNN instance.
+
+        Args:
+            nstate: number of hidden units.
+            nin: number of input units.
+            nout: number of output units.
+            activation: actication function for hidden layer.
+            w_init: weight initializer for RNN model weights.
+        """
+        self.num_inputs = nin
+        self.num_outputs = nout
+        self.nstate = nstate
+        self.activation = activation
+
+        # Hidden layer parameters
+        self.w_xh = TrainVar(w_init((self.num_inputs, self.nstate)))
+        self.w_hh = TrainVar(w_init((self.nstate, self.nstate)))
+        self.b_h = TrainVar(jn.zeros(self.nstate))
+
+        self.output_layer = Linear(self.nstate, self.num_outputs)
+
+    def __call__(self, inputs: JaxArray, only_return_final=False) -> JaxArray:
```

Comment:
suggest adding a get_initial_state method and optional initial_state argument here

Reply:
I added an optional initial_state argument to the call() method. 

Can you clarify what the get_initial_state() method would do, considering that the state is initialized during every call() (unless explicitly passed in through the optional argument)?

Reply:
there are two reasons to have a `get_initial_state`: One, the caller wants to know if this layer is recurrent, without checking for some general instance type.  Two, the caller wants to know the shapes etc of the state, without running `__call__`.  This is useful for many reasons, like creating buffers for storing state.

Reply:
Just to clarify, does `get_init_state` really act like a `create_init_state`? Or is there an `init_state` stored inside the instance?

Reply:
no; it's a purely functional thing that returns some arrays.

Reply:
As far as I understood from some of the Keras code, `get_initial_state` simply returns zero array of appropriate shape (ex: https://github.com/tensorflow/tensorflow/blob/fcc4b966f1265f466e82617020af93670141b009/tensorflow/python/keras/layers/recurrent.py#L1948 )

It's still not very clear how useful it is.
Could you point us to some example of how it's actually used (either in Tensorflow or any other framework)?

To know shape of the state it would be better to just call `rnn_cell_layer.nstate` or maybe have helper method `get_state_shape`.
Using `get_initial_state` as a way to determine whether layer is RNN seems like a little weird. I don't see how `getattr(layer, 'get_initial_state')` is better than `isinstance(layer, RNNCell)`. If there is a need to determine whether layer is RNN cell, I think it's better just to make all RNN cells to inherit from some base class and do `isinstance` check.